### DISCUSSION TITLE
Boolen operation (difference/union) between plane and ellipsoid

### DISCUSSION THREAD

**Discussion by mhueppe** [pyvista/pyvista#4893] [Q&A]
## Description

I want to create a pothole (ellipsoid) in a street (plane). After some testing, I realized that the boolean union is probably the one to use (as I want to combine both of the meshes however do not want the street to cover up the pothole).
When I try this using a Sphere it works perfectly fine.
However, I am using a ParameticEllipsoid. There I constantly get the error "vtkMath::Jacobi: Error extracting eigenfunctions" and the wanted effect only works with the boolean difference instead of the union. I suppose that it has to do something with the vtk/py and its geometric object creation.

As I said taking the boolean difference returns good results it does so only rarely (crashing with too many objects). 

Is there a way to avoid this? My idea was to extract the points of the meshes (NumPy array) and calculate the union with them however I wasn't able to.
Another idea of mine was to clip the plane (street) by a surface (potholes) and then add a halved ellipsoid however this results in cracks in the surface. 

## Example Code

```
def add_pothole():
    street_mesh = pv.Plane(i_size=100, j_size=100, i_resolution=100, j_resolution=100)

    points = street_mesh.points
    for i in range(4):
        pothole_mesh = pv.ParametricEllipsoid(10,5,3)
        # alternatively to get the half ellipsoid: (but boolean difference seemed to only work with the closed one
        # pothole_mesh = pv.ParametricEllipsoid(10,5,3, max_v=pi /2)
        pothole_mesh.translate(choice(points))

        # "add" the pothole to the street
        street_mesh = pothole_mesh.boolean_difference(street_mesh.delaunay_2d())

    street_mesh.plot()
```

**Comment by akaszynski** [comment]
Sorry for the late reply.

Boolean filters are quite flaky within VTK.  I think `pv.ParametricEllipsoid` may not be returning a manifold surface.  Try using [clean](https://docs.pyvista.org/core/filters.html?highlight=clean#pyvista.PolyDataFilters.clean) to merge any potential inconsistencies within the mesh, and if that doesn't work, you might try using different surfaces.

**Comment by GlennWSo** [comment]
Here a few tips:

ParametricEllipsoid does not have consitent normals by default. From what i can tell, most of them are pointing inwards, exept one normal. This incorrect. To fix this u can use compute_normals like so:
```
pothole_mesh = pv.ParametricEllipsoid(
            10, 5, 3, clean=True, u_res=res, v_res=res, w_res=res
        ).compute_normals(flip_normals=True)
```
i you can i would dial down the resultion of the ellipsiod, since the boolean operations can be expensive.
the boolean operations in pyvista not just buggy, but also slow.

Some faster and more robust options
warning: incoming self plug
if you are using linux you may want to try: rscad
rscad has boolean operations that act on and return pv.PolyData

another good options for booleans in python is pymadcad 

using the rscad i was able to get this result in less 1.0 second:
![potholes](https://user-images.githubusercontent.com/107837123/222482198-3eee5e82-1cd2-4b98-89e6-598e01adca40.jpg)


```python
from time import time


import pyvista as pv
import numpy as np
from rscad.boolean.polydata import diff


def add_pothole():
    t0 = time()
    street_mesh = pv.Plane(
        i_size=100, j_size=100, i_resolution=100, j_resolution=100
    ).triangulate()
    # street_mesh.flip_normals()

    points = np.array(
        [
            [20, 30, 0],
            [50, 50, 0],
            [60, 30, 0],
            [20, 60, 0],
        ]
    ) - (50, 50, 0)
    for i in range(4):
        res = 40
        pothole_mesh = pv.ParametricEllipsoid(
            10, 5, 3, clean=True, u_res=res, v_res=res, w_res=res
        ).compute_normals(flip_normals=True)

        # ensure normals look normal ;) i.e thay are consistent and pointing outwards
        # pothole_mesh.plot_normals()

        # alternatively to get the half ellipsoid: (but boolean difference seemed to only work with the closed one
        # pothole_mesh = pv.ParametricEllipsoid(10,5,3, max_v=pi /2)
        pothole_mesh.translate(points[i], inplace=True)

        # "add" the pothole to the street
        street_mesh = diff(street_mesh, pothole_mesh)
        # street_mesh = street_mesh.boolean_difference(pothole_mesh)

    t = time() - t0
    print(f"time: {t} s")
    street_mesh.plot()


if __name__ == "__main__":
    add_pothole()

```
